### PR TITLE
feat(#1156): provide test api generator

### PR DESCRIPTION
**Pull request by tschlat**
I have added code for the test API generator plus documentation. The generator has been implemented having XML tests in mind. Therefore, usage within Java DSL is a little cumbersome and needs a rework. We propose to add this change as work in progress. It is quite isolated.

I have added #1161 and #1163 to finish this up.

### CODE REVIEW

File: core/citrus-api/src/main/java/org/citrusframework/testapi/GeneratedApi.java [hunk 1]
```diff
@@ -0,0 +1,26 @@
+package org.citrusframework.testapi;
+
+import java.util.Map;
+
+public interface GeneratedApi {
```
Comment (bbortt) [inline core/citrus-api/src/main/java/org/citrusframework/testapi/GeneratedApi.java:5]: I think a javadoc on these API interfaces would be well-invested.

File: src/manual/testapi.adoc
Comment (bbortt) [inline src/manual/testapi.adoc:1]: this file must be included somewhere, right? I would suggest to add it after ["5. Running tests in Java"](https://github.com/citrusframework/citrus/blob/main/src/manual/index.adoc?plain=1#L31).

File: src/manual/testapi.adoc [hunk 2]
```diff
@@ -0,0 +1,380 @@
+[[testapi]]
+= Test API Generation
+
+== OpenAPI: A Standard for API Description
+OpenAPI, formerly known as Swagger, is a widely adopted standard for describing RESTful APIs. It provides a
```
Comment (bbortt) [inline src/manual/testapi.adoc:5]: I usually do one sentence per line in `*.adoc`. that makes it much more comparable in later pull-requests. would you mind giving it a refactoring? 👼🏼 sorry...

File: src/manual/testapi.adoc [hunk 3]
```diff
@@ -0,0 +1,380 @@
...
+language-agnostic interface to define the structure of APIs in a human-readable format, typically using YAML
```
Comment (bbortt) [inline src/manual/testapi.adoc:6]:
```suggestion
language-agnostic interface to define the structure of APIs in a human- and at the same time machine-readable format, typically using YAML
```

File: src/manual/testapi.adoc [hunk 4]
```diff
@@ -0,0 +1,380 @@
...
+or JSON. OpenAPI specifications serve as a contract for your API, detailing endpoints, request/response formats,
+authentication methods, and more.
+
+OpenAPI specifications are commonly used to generate various artifacts such as client libraries, server stubs, and
+documentation. By capturing the structure and behavior of APIs in a machine-readable format, OpenAPI enables
+seamless code generation, saving time and ensuring consistency across different implementations.
```
Comment (bbortt) [inline src/manual/testapi.adoc:12]:
```suggestion
```

you've already said that imo.

File: src/manual/testapi.adoc [hunk 5]
```diff
@@ -0,0 +1,380 @@
...
+`Citrus OpenAPI Test API Generator` provides the following features:
+
+*   generation of a Test API
+** from OpenAPI Specification
+**  [TODO #1163] from WSDL via an intermediate step that generates a "light" OpenApi specification from a WSDL
+*   integration into Citrus XML test cases
+**  integration into XML editors via provided XSD
+***  schema validation
+***  auto completion
+*   integration into Citrus Java test cases
```
Comment (bbortt) [inline src/manual/testapi.adoc:32]:
```suggestion
*   integration into <<run-java,Citrus Java test cases>>
```

File: src/manual/testapi.adoc [hunk 6]
```diff
@@ -0,0 +1,380 @@
...
+seamlessly integrates into Citrus XML or Java tests, ensuring a cohesive testing experience.
+
+== What does it offer?
+
+`Citrus OpenAPI Test API Generator` provides the following features:
+
+*   generation of a Test API
+** from OpenAPI Specification
+**  [TODO #1163] from WSDL via an intermediate step that generates a "light" OpenApi specification from a WSDL
+*   integration into Citrus XML test cases
```
Comment (bbortt) [inline src/manual/testapi.adoc:28]:
```suggestion
*   integration into <<run-xml-tests,Citrus XML test cases>>
```

File: src/manual/testapi.adoc [hunk 7]
```diff
@@ -0,0 +1,380 @@
...
+***  schema validation
+***  auto completion
+*   integration into Citrus Java test cases
+
+=== How Does it Work?
+
+The `Citrus OpenAPI Test API Generator` leverages the link:https://github.com/swagger-api/swagger-codegen/tree/master[OpenAPI Code Generator]
+to generate code, but provides custom templates tailored for seamless integration within the Citrus framework.
+
+During a build process, code generation is carried out by a Maven plugin. While the standard generator plugin,
```
Comment (bbortt) [inline src/manual/testapi.adoc:39]:
```suggestion
During a build process, code generation is carried out by a Maven or Gradle plugin. While the standard generator plugin,
```

File: src/manual/testapi.adoc [hunk 8]
```diff
@@ -0,0 +1,380 @@
...
+`org.openapitools:openapi-generator-maven-plugin`, can be employed for this purpose, configuring it can be
+cumbersome, especially when dealing with multiple APIs. To address this challenge, Citrus offers its adaptation
+of this standard generator Maven plugin. This `Citrus OpenAPI Generator Plugin` simplifies the configuration of test API
```
Comment (bbortt) [inline src/manual/testapi.adoc:42]:
```suggestion
of this standard plugin. This `Citrus OpenAPI Generator Plugin` simplifies the configuration of test API
```

File: src/manual/testapi.adoc [hunk 9]
```diff
@@ -0,0 +1,380 @@
...
+
+| `schemaFolder`                   | `citrus.test.api.generator.schema.folder`                     | Location for the generated XSD schemas                           | `schema/xsd/%VERSION%`
+| `resourceFolder`                 | `citrus.test.api.generator.resource.folder`                   | Location to which resources are generated                        | `generated-resources`
+| `sourceFolder`                   | `citrus.test.api.generator.source.folder`                     | Location to which sources are generated                          | `generated-sources`
+| `metaInfFolder`                  | `citrus.test.api.generator.meta.inf.folder`                   | Location to which spring meta files are generated/updated        | `target/generated-test-resources/META-INF`
+| `generateSpringIntegrationFiles` | `citrus.test.api.generator.generate.spring.integration.files` | Specifies whether spring integration files should be generated   | `true`
+| Nested api element               |                                                               |                                                                   |
```
Comment (bbortt) [inline src/manual/testapi.adoc:198]:
```suggestion
| Nested `<api />` element               |                                                               |                                                                   |
```

File: src/manual/testapi.adoc [hunk 10]
```diff
@@ -0,0 +1,380 @@
...
+
+Note: `%PREFIX%` and `%VERSION%` are placeholders that will be replaced by their specific values as configured. The plugin performs a conversion to lowercase for `PREFIX` used in package names and in `targetXmlnsNamespace`.
+
+=== Running the generator
+
+To run the generator, execute the following command in your project directory:
+
+[source,bash]
+----
+mvn citrus-test-api-generator-maven-plugin:create-test-api
```
Comment (bbortt) [inline src/manual/testapi.adoc:220]: are you sure this works? mustn't it be fully qualified?

File: src/manual/testapi.adoc [hunk 11]
```diff
@@ -0,0 +1,380 @@
...
+----
+
+
+This command will generate the classes and XSD files as configured for your APIs in the specified locations.
```
Comment (bbortt) [inline src/manual/testapi.adoc:224]:
```suggestion
This command will generate the classes and XSD files as configured for your APIs in the specified locations.
```

File: src/manual/testapi.adoc [hunk 12]
```diff
@@ -0,0 +1,380 @@
...
+----
+
+=== Sample usage
+
+To utilize the test API in XML, it's necessary to import the respective namespace. Once imported, requests
+can be directly employed as actions, as illustrated in the sample below. Further examples can be found here
+`org.citrusframework.openapi.generator.GeneratedApiIT`.
+
+
+
```
Comment (bbortt) [inline src/manual/testapi.adoc:314]:
```suggestion

```

File: src/manual/testapi.adoc [hunk 13]
```diff
@@ -0,0 +1,380 @@
...
+                    <petstore:json-path value="Snoopy" expression="$.name" />
+                    <petstore:json-path value="12" expression="$.id" />
+                </petstore:response>
+            </petstore:getPetByIdRequest>
+        </actions>
+    </testcase>
+</spring:beans>
+----
+
+
```
Comment (bbortt) [inline src/manual/testapi.adoc:343]:
```suggestion

```

File: src/manual/testapi.adoc [hunk 14]
```diff
@@ -0,0 +1,380 @@
...
+To utilize the test API in Java, it's necessary to import the API configuration, that provides the respective
+request actions. The request to test can then be autowired, configured and autowired, as illustrated in the sample below.
```
Comment (bbortt) [inline src/manual/testapi.adoc:345]:
```suggestion
request actions. The request to test can then be configured and autowired, as illustrated in the sample below.
```

File: test-api-generator/citrus-test-api-generator-core/pom.xml [hunk 15]
```diff
@@ -0,0 +1,212 @@
...
+        <dependency>
+            <groupId>org.citrusframework</groupId>
+            <artifactId>citrus-junit5</artifactId>
+            <version>${project.version}</version>
+            <scope>test</scope>
+        </dependency>
+        <dependency>
+            <groupId>org.springframework.boot</groupId>
+            <artifactId>spring-boot-test</artifactId>
+            <version>3.2.5</version>
```
Comment (bbortt) [inline test-api-generator/citrus-test-api-generator-core/pom.xml:72]: I think this should use [the managed spring version](https://github.com/citrusframework/citrus/blob/e05ec958626719c3f3eca8eecc2bac70f2ecf4b6/pom.xml#L256):

```suggestion
            <version>${spring.version}</version>
```

File: test-api-generator/citrus-test-api-generator-core/pom.xml [hunk 16]
```diff
@@ -0,0 +1,212 @@
...
+        <dependency>
+            <groupId>org.citrusframework</groupId>
+            <artifactId>citrus-validation-json</artifactId>
+            <version>${project.version}</version>
+            <scope>test</scope>
+        </dependency>
+        <dependency>
+            <groupId>org.junit.jupiter</groupId>
+            <artifactId>junit-jupiter-params</artifactId>
+            <version>5.10.2</version>
```
Comment (bbortt) [inline test-api-generator/citrus-test-api-generator-core/pom.xml:89]: use the [managed junit jupiter version](https://github.com/citrusframework/citrus/blob/e05ec958626719c3f3eca8eecc2bac70f2ecf4b6/pom.xml#L238):

```suggestion
            <version>${junit.jupiter.version}</version>
```

File: test-api-generator/citrus-test-api-generator-core/src/main/java/org/citrusframework/openapi/generator/JavaCitrusCodegen.java [hunk 17]
```diff
@@ -0,0 +1,345 @@
...
+import java.util.Set;
+import org.openapitools.codegen.CodegenType;
+import org.openapitools.codegen.SupportingFile;
+import org.openapitools.codegen.languages.AbstractJavaCodegen;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * @author Thorsten Schlathoelter
+ */
```
Comment (bbortt) [inline test-api-generator/citrus-test-api-generator-core/src/main/java/org/citrusframework/openapi/generator/JavaCitrusCodegen.java:22]:
```suggestion
```

personally, I don't like these `@author` comments.. that's all part of the git history. but maybe you (or @christophd) have other opinions on that matter.

anyway, all your files are missing the copyright comment on top of them. I think this is especially important for java source files, not necessarily for other types.

File: test-api-generator/citrus-test-api-generator-core/src/test/resources/apis/multiparttest-rest-resource.yaml [hunk 18]
```diff
@@ -0,0 +1,249 @@
...
+info:
+    title: multiparttest API
+    version: 2.0.0
+    description: |
+        The service to test mutlipart
+    x-citrus-app: MPT
+    x-citrus-api-name: multiparttest-rest-resource
+    contact:
+        name: IT-Services-CI TAuBE
+        email: IT-Serv-CI-ETAdl@post.ch
```
Comment (bbortt) [inline test-api-generator/citrus-test-api-generator-core/src/test/resources/apis/multiparttest-rest-resource.yaml:11]: please use random name, email and url 😉

File: test-api-generator/citrus-test-api-generator-maven-plugin/pom.xml [hunk 19]
```diff
@@ -0,0 +1,124 @@
...
+        <dependency>
+            <groupId>io.swagger.core.v3</groupId>
+            <artifactId>swagger-models-jakarta</artifactId>
+            <version>${swagger.version}</version>
+        </dependency>
+
+        <dependency>
+            <groupId>commons-io</groupId>
+            <artifactId>commons-io</artifactId>
+            <version>2.16.1</version>
```
Comment (bbortt) [inline test-api-generator/citrus-test-api-generator-maven-plugin/pom.xml:62]: https://github.com/citrusframework/citrus/blob/e05ec958626719c3f3eca8eecc2bac70f2ecf4b6/pom.xml#L203:

```suggestion
            <version>${commons.io.version}</version>
```

File: test-api-generator/citrus-test-api-generator-core/pom.xml [hunk 20]
```diff
@@ -0,0 +1,212 @@
...
+        </dependency>
+        <dependency>
+            <groupId>org.citrusframework</groupId>
+            <artifactId>citrus-api</artifactId>
+            <version>4.3.0-SNAPSHOT</version>
+        </dependency>
+        <dependency>
+            <groupId>org.citrusframework</groupId>
+            <artifactId>citrus-http</artifactId>
+            <version>4.3.0-SNAPSHOT</version>
```
Comment (bbortt) [inline test-api-generator/citrus-test-api-generator-core/pom.xml:48]: I think these for (`citrus-api` too) should be `${project.version}`, right?

File: test-api-generator/citrus-test-api-generator-maven-plugin/pom.xml [hunk 21]
```diff
@@ -0,0 +1,124 @@
...
+            <groupId>org.apache.maven.plugin-testing</groupId>
+            <artifactId>maven-plugin-testing-harness</artifactId>
+            <version>3.3.0</version>
+            <scope>test</scope>
+        </dependency>
+
+        <dependency>
+            <groupId>org.openapitools</groupId>
+            <artifactId>openapi-generator-maven-plugin</artifactId>
+            <version>7.4.0</version>
```
Comment (bbortt) [inline test-api-generator/citrus-test-api-generator-maven-plugin/pom.xml:110]: just a side note: `7.5.0` is out already, I think.

File: test-api-generator/pom.xml [hunk 22]
```diff
@@ -0,0 +1,38 @@
...
+    <dependencies>
+        <dependency>
+            <groupId>org.openapitools</groupId>
+            <artifactId>openapi-generator</artifactId>
+            <version>7.5.0</version>
+        </dependency>
+        <dependency>
+            <groupId>org.junit.jupiter</groupId>
+            <artifactId>junit-jupiter-engine</artifactId>
+            <version>5.10.2</version> <!-- Use the latest version -->
```
Comment (bbortt) [inline test-api-generator/pom.xml:27]: why not the managed version?

File: test-api-generator/pom.xml [hunk 23]
```diff
@@ -0,0 +1,38 @@
...
+    <artifactId>citrus-test-api-generator</artifactId>
+    <name>Citrus :: Test API Generator</name>
+    <description>Citrus Test API Generator</description>
+    <packaging>pom</packaging>
+
+    <dependencies>
+        <dependency>
+            <groupId>org.openapitools</groupId>
+            <artifactId>openapi-generator</artifactId>
+            <version>7.5.0</version>
```
Comment (bbortt) [inline test-api-generator/pom.xml:22]: pretty sure these version is always the same as for the maven plugin (https://github.com/citrusframework/citrus/pull/1160/files#diff-a85c5bf4e7da63f6dbd8421e55cf9fe495a2136897f394368514bca2d5daadfeR110)). would probably be worht adding to the global managed versions?

**Review by bbortt — CHANGES_REQUESTED** [review]
huge @tschlat! I won't be nitpicking on this 🐤 but left some remarks.

File: src/manual/testapi.adoc
Comment (tschlat) [reply src/manual/testapi.adoc:1]: Moved it to `connector-openapi.adoc` and embdedded it into that context.

File: src/manual/testapi.adoc (hunk 2 above)
Comment (tschlat) [reply src/manual/testapi.adoc:5]: It's currently not really done that way but i changed my code respectively.

File: src/manual/testapi.adoc (hunk 4 above)
Comment (tschlat) [reply src/manual/testapi.adoc:12]: Same here.

File: test-api-generator/citrus-test-api-generator-maven-plugin/pom.xml (hunk 19 above)
Comment (tschlat) [reply test-api-generator/citrus-test-api-generator-maven-plugin/pom.xml:62]: I replaced all explicit versions with properties, which are defined on root pom. Added the missing ones.

File: test-api-generator/citrus-test-api-generator-core/src/main/java/org/citrusframework/openapi/generator/JavaCitrusCodegen.java (hunk 17 above)
Comment (christophd) [reply test-api-generator/citrus-test-api-generator-core/src/main/java/org/citrusframework/openapi/generator/JavaCitrusCodegen.java:22]: +1 on skipping the `@author` information as this is really in the github history.

Adding the copyright license information to all source files is more important

**Comment by christophd** [comment]
@tschlat awesome work! please bear with me as I need some more time to have a closer look

**Comment by tschlat** [comment]
Closed in favor of: https://github.com/citrusframework/citrus/pull/1224

**Comment by tschlat** [comment]
Closed in favor of: https://github.com/citrusframework/citrus/pull/1224

**Pull request closed without merging**